server/storage/schema/schema.go

```diff
@@ -69,6 +75,12 @@ func UnsafeMigrate(lg *zap.Logger, tx backend.BatchTx, target semver.Version) er
 	if err != nil {
 		return fmt.Errorf("cannot create migration plan: %w", err)
 	}
+	if target.LessThan(current) {
+		minVersion := w.MinimalEtcdVersion()
+		if minVersion != nil && target.LessThan(*minVersion) {
+			return fmt.Errorf("cannot downgrade storage, WAL contains newer entries")
```

Comment:
I see this error message in downgrade test occasionally. See example below,
```
{"level":"error","ts":"2024-12-18T09:53:14.334189Z","caller":"version/monitor.go:120","msg":"failed to update storage version","cluster-version":"3.5.0","error":"cannot downgrade storage, WAL contains newer entries","stacktrace":"go.etcd.io/etcd/server/v3/etcdserver/version.(*Monitor).UpdateStorageVersionIfNeeded\n\tgo.etcd.io/etcd/server/v3/etcdserver/version/monitor.go:120\ngo.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).monitorStorageVersion\n\tgo.etcd.io/etcd/server/v3/etcdserver/server.go:2296\ngo.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).GoAttach.func1\n\tgo.etcd.io/etcd/server/v3/etcdserver/server.go:2477"}
```

The reason is it read the version "3.6.0" included in `ClusterVersionSet` request,

```
2	         9	norm	header:<ID:7587883512398433796 > cluster_version_set:<ver:"3.6.0" >
```

See link below,
https://github.com/etcd-io/etcd/blob/0966b4d68ec9ff443a43431eae71305796e3200f/server/storage/wal/version.go#L115-L120

When etcd applies the `ClusterVersionSet` request, it will forcibly create a snapshot, but there (setting cluster version, and creating snapshot) are an atomic operation. If etcd happens to try to update the storage version in between, then you will see the error message. It won't do any harm, but it may cause confusion to users.
https://github.com/etcd-io/etcd/blob/0966b4d68ec9ff443a43431eae71305796e3200f/server/etcdserver/apply/apply.go#L417-L427